### PR TITLE
Changes after rebase Mar 2022

### PR DESCRIPTION
**Pull request by 3vcloud**
Fixed issues around send/recv ping (don't need to send a response ping where defined)
Fixed issues around array_reserve after recent change to definition
Warning related changes and compilation issues

### CODE REVIEW

File: code/client/CMakeLists.txt [hunk 1]
```diff
@@ -85,7 +85,7 @@ target_include_directories(
   client
 
   PUBLIC
-  ${CMAKE_SOURCE_DIR}/include
+  ../../include
```
Comment (reduf) [inline code/client/CMakeLists.txt:88]: Why is that necessary?

File: code/portal/login.c [hunk 2]
```diff
@@ -32,7 +32,7 @@ static struct str s_from_c_str(const char *ptr)
 static int s_cmp(struct str *str1, struct str *str2)
 {
     if (str1->len != str2->len)
-        return str1->len - str2->len;
+        return (int)str1->len - (int)str2->len;
```
Comment (reduf) [inline code/portal/login.c:35]: This is my bad, but it's not safe to cast to a `int`.

I guess, in this case it would be better to simply do:
```c
if (str1->len < str2->len)
    return -1;
else if (str1->len > str2->len)
    return 1;
else
    return memcmp(str1->ptr, str2->ptr, str1->len);
```

File: code/common/array.c [hunk 3]
```diff
@@ -74,13 +74,12 @@ _array_resize(array_void_t *a, size_t size, const size_t elem_size)
 
     if (a->capacity < size) {
         int ret;
-        if ((ret = _array_grow_to(a, size, elem_size)) != 0) {
+        if ((ret = _array_reserve(a, size, elem_size)) != 1)
```
Comment (reduf) [inline code/common/array.c:77]: Why changing `_array_grow_to` to `_array_reserve`?

That looks wrong, but what was your case?

File: code/portal/socket.c [hunk 4]
```diff
@@ -25,7 +25,7 @@ int recv_to_buffer(SOCKET fd, array_uint8_t *buffer)
     const size_t BUFFER_SIZE = 1024;
     uint8_t *ptr = array_push(buffer, BUFFER_SIZE);
 
-    if ((ret = recv(fd, (char *)ptr, BUFFER_SIZE, 0)) <= 0) {
+    if ((ret = recv(fd, (char *)ptr, (int)BUFFER_SIZE, 0)) <= 0) {
```
[the remark this answers is not in the document]
Comment (3vcloud) [reply code/portal/socket.c:28]: https://docs.microsoft.com/en-us/windows/win32/api/winsock/nf-winsock-recv

File: code/common/array.c (hunk 3 above)
Comment (3vcloud) [reply code/common/array.c:77]: So resize and reserve were broken since your recent changes; I decided to use std::string as the basic logic, so when resizing we grow the array and malloc. I think the initial problem was that `array_resize` wasn't actually resizing, or setting the size variable

File: code/portal/socket.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (reduf) [reply code/portal/socket.c:28]: That's rather annoying, Unix version uses [size_t](https://man7.org/linux/man-pages/man2/recv.2.html).

File: code/common/array.c (hunk 3 above)
Comment (reduf) [reply code/common/array.c:77]: So, `array_resize` was indeed not following `std::string` (or `std::vector`) which I wanted to change recently. Likely where I introduced the bug.

But, `array_grow_to` should effectively be the same as `array_reserve`, pretty much the internal version. There is still one inaccuracy between `array_reserve` and `std::vector::reserve`. `std::vector::reserve` reserve places for N elements, instead of N additional elements. So, if the array already have elements, adding N elements should be reserved as `a.reserve(a.size() + N)`. Maybe we should change it to.

Anyhow, the `_array_grow_to` seems fine in this case, do you have the case that broke it? I'm slowly adding unit tests, because this project is severely lacking unit tests.

Comment (3vcloud) [reply code/common/array.c:77]: I think the first issue to do with use of arrays was finding the ip address of the auth server before any authentication is made. Also had issues with the number of available bag slots and another issue around auth server handlers

Comment (3vcloud) [reply code/common/array.c:77]: Also I think `std::vector::resize` also frees memory if the required size is less than current - didn't want to enforce that because that might cause a whole heap of more problems

File: code/portal/login.c (hunk 2 above)
Comment (reduf) [reply code/portal/login.c:35]: I fixed it.

File: code/portal/CMakeLists.txt [hunk 5]
```diff
@@ -23,6 +23,8 @@ set_source_files_properties(
 add_library(
   portal
 
+STATIC
```
[the remark this answers is not in the document]
Comment (3vcloud) [reply code/portal/CMakeLists.txt:26]: If we're building a plugin that has a dependency on client, the plugin needs to be a shared lib but everything else doesn't - wouldn't setting this flag in the plugin's cmake file cause issues?

[the remark this answers is not in the document]
Comment (reduf) [reply code/portal/CMakeLists.txt:26]: Right, so if something needs to be a plugin, the it should be the job of this target to set itself as `SHARED`. Every plugins should set itself as shared.

I will likely at some point add a "cmake" folder that we include useful functions for plugins. (e.g., `add_headquarter_plugin`)

[the remark this answers is not in the document]
Comment (3vcloud) [reply code/portal/CMakeLists.txt:26]: any way to set this via `set_target_properties` ?

[the remark this answers is not in the document]
Comment (reduf) [reply code/portal/CMakeLists.txt:26]: I'm sure there is a way, no idea if it's `set_target_properties`.

File: code/client/game.c [hunk 6]
```diff
@@ -5,6 +5,7 @@
 
 static void HandlePingRequest(Connection *conn, size_t psize, Packet *packet)
 {
+    (psize);
```
[the remark this answers is not in the document]
Comment (reduf) [reply code/client/game.c:8]: This problem was addressed correctly. The issue was simply due to format of packet changing since it was implemented.

**Comment by reduf** [comment]
Some commits were cherry picked, other issues were solved with alternative solutions. All problems should have been addressed.
